### PR TITLE
Adding an upload endpoint for CUIT to hit to upload data

### PR DESCRIPTION
**Pull request by mjp2220**
Finally wrote the endpoint to allow CUIT to upload data to Density.

Can write out the specifics of the endpoint once I get a bit of code review.

@alanhdu hoping you can review this

### CODE REVIEW

File: density/db/db.py [hunk 1]
```diff
@@ -210,3 +211,39 @@ def get_uni_for_code(cursor, code):
...
+  for key in data:
+    query = """INSERT INTO {table_name}
+             (dump_time, group_id, group_name, parent_id, parent_name, client_count)
+             VALUES (%s, %s, %s, %s, %s, %s);""".format(table_name=TABLE_NAME)
+    try:
+      cursor.execute(query, (date, int(key), data[key]['name'], int(data[key]['parent_id']), PARENTS[data[key]['parent_id']], int(data[key]['client_count'])))
+    except:
+      db_failed += 1
+
+  return db_failed
```
Comment (alanhdu) [inline density/db/db.py]: Should this just be a `bool` instead?

File: density/db/db.py [hunk 2]
```diff
@@ -210,3 +211,39 @@ def get_uni_for_code(cursor, code):
...
+  # It's easy with list comprehensions to batch them, but hard to escape
+  # single quotes that may appear in group names.
+  for key in data:
+    query = """INSERT INTO {table_name}
+             (dump_time, group_id, group_name, parent_id, parent_name, client_count)
+             VALUES (%s, %s, %s, %s, %s, %s);""".format(table_name=TABLE_NAME)
+    try:
+      cursor.execute(query, (date, int(key), data[key]['name'], int(data[key]['parent_id']), PARENTS[data[key]['parent_id']], int(data[key]['client_count'])))
+    except:
+      db_failed += 1
```
Comment (alanhdu) [inline density/db/db.py]: Why are we swallowing a possible failed insertion error? why not just bubble it up and fail the request if necessary?

If we are going to swallow the error, we should at least log it for our own sanity.

File: density/db/db.py [hunk 3]
```diff
@@ -210,3 +211,39 @@ def get_uni_for_code(cursor, code):
...
+             VALUES %s;""".format(table_name=TABLE_NAME)
+
+  # Set the time
+  date = datetime.now().replace(second=0, microsecond=0)
+
+  db_failed = 0
+
+  # NOTE: We tried batching these but ran into some problems escaping quotes.
+  # It's easy with list comprehensions to batch them, but hard to escape
+  # single quotes that may appear in group names.
```
Comment (alanhdu) [inline density/db/db.py]: Could you be more specific about the problem?

(Also, I believe `cursor.executemany` should work here)

File: density/density.py [hunk 4]
```diff
@@ -493,5 +493,27 @@ def map():
...
+    if secret_key != app.config['UPLOAD_KEY']:
+        return 'Please include a valid API key.', 401
+
+    json_data = request.get_json()
+    if 'data' not in json_data:
+        return 'Please include data to upload.', 400
+
+    density_data = json_data['data']
+
+    db_failed = db.insert_density_data(g.cursor, density_data)
```
Comment (alanhdu) [inline density/density.py]: Maybe we should check for other conditions? Like make sure that the data contains all rooms we support?

File: density/density.py [hunk 5]
```diff
@@ -493,5 +493,27 @@ def map():
...
     return render_template('map.html', locations=locations)
 
+@app.route('/upload', methods = ['POST'])
+def upload():
+    """ Accept POST requests from CUIT to add new data to the server """
+
+    # This is stored in local settings and is the way we verify uploads.
+    secret_key = request.args['key']
+    if secret_key != app.config['UPLOAD_KEY']:
+        return 'Please include a valid API key.', 401
```
Comment (alanhdu) [inline density/density.py]: I feel like this should really be an HTTP header, but this works too.

**Comment by alanhdu** [comment]
Also needs to pass Travis (I suspect it's just `flake8` stuff).

File: density/density.py (hunk 5 above)
Comment (mjp2220) [inline density/density.py]: Gonna keep it as a key, I think it should be easier for CUIT to do it as a URL param.

**Comment by mjp2220** [comment]
Fixed the other stuff - used `cursor.executemany`, that fixed the issue. Plus cleaned up some of the error handling, and got the flake8 working.

@alanhdu can you handle merging/deploying this? I want to start working with CUIT to get Density functional again.

File: density/density.py [hunk 6]
```diff
@@ -508,12 +508,13 @@ def upload():
...
 
-    db_failed = db.insert_density_data(g.cursor, density_data)
-    if db_failed == 0:
+    db_success = db.insert_density_data(g.cursor, density_data)
+    if db_success:
         return 'Data successfully uploaded.', 200
     else:
-        return '%d rows were not inserted. Please try again or contact someone in ADI.' % db_failed, 500
+        return 'At least one of the records was not inserted. \
+            Please contact someone in ADI for more details.', 500
```
Comment (alanhdu) [inline density/density.py]: So I personally think that `db.insert_density_data` should just throw an error, and we should have a generic Flask error handler to tell people to contact ADI.

**Comment by alanhdu** [comment]
@mjp2220: So I'm running into https://github.com/conda/conda/issues/5004 when I try to deploy as-is -- I'm working on debugging the upstream issue. Should I wait for that (and possibly for you to update the PR), or should I just rip out and rewrite our deployment mechanism?

File: density/db/db.py [hunk 7]
```diff
@@ -212,38 +212,43 @@ def get_uni_for_code(cursor, code):
...
+    data = [(date, int(key), data[key]['name'],
+            int(data[key]['parent_id']), PARENTS[data[key]['parent_id']],
+            int(data[key]['client_count'])) for key in data]
+
+    try:
+        cursor.executemany(query, data)
+    except Exception as e:
+        # At least log the error for our own sanity. But, we still
+        # want to insert as many records as we can, so we won't break.
+        print(e)
```
Comment (alanhdu) [inline density/db/db.py]: Can we use Flask's logger here instead of printing?

**Comment by mjp2220** [comment]
@alanhdu feel free to rip it out. Also, I didn't find any examples of the logging.

Hoping we can get this deployed today or tomorrow, if we can. Got several emails in my inbox about it, hoping to finally respond to them.

**Comment by alanhdu** [comment]
@mjp2220: Merged and deployed, although I added an extra commit on top of it (046c5836c8d827a01870ee9612bbd5669d82f750).

One question though: where did the `PARENTS` dict come from?

**Pull request merged**